src/Hooks/StrictEqualityHooks.php

```diff
@@ -121,42 +134,184 @@ private static function isEqualOrdered(Atomic $first_type, Atomic $second_type):
+
+    private static function isNotEqualOrdered(Atomic $first_type, Atomic $second_type): bool
+    {
+        // identical at the moment
+        return self::isEqualOrdered($first_type, $second_type);
+    }
+
+    private static function isUnionCompatibleType(array $left_type_atomics, array $right_type_atomics, string $expr_class): bool {
+        if ($expr_class === Equal::class) {
+            //This is just a trick to avoid handling every way
+            return self::isUnionEqualOrdered($left_type_atomics, $right_type_atomics) || self::isUnionEqualOrdered($right_type_atomics, $left_type_atomics);
+        } else {
+            return self::isUnionNotEqualOrdered($left_type_atomics, $right_type_atomics) || self::isUnionNotEqualOrdered($right_type_atomics, $left_type_atomics);
+        }
+    }
+
+    private static function isUnionStringEqualOrdered(array $first_types, array $second_types): bool
+    {
+        foreach ($first_types as $atomic_type) {
+            if ($atomic_type instanceof Atomic\TNonEmptyString) {
+                $with_null = true;
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TString) {
+                $with_null = false;
+                continue;
+            }
+
+            // only single, not-union string is fixable at the moment
+            return false;
+        }
+
+        foreach ($second_types as $atomic_type) {
+            if ($atomic_type instanceof Atomic\TString) {
+                continue;
+            }
+
+            if ($with_null === true && $atomic_type instanceof Atomic\TNull) {
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TArray) {
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TKeyedArray) {
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TList) {
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TIterable) {
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TObject) {
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TNamedObject) {
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TCallable) {
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TResource) {
+                continue;
+            }
+
+            if ($atomic_type instanceof Atomic\TClosedResource) {
+                continue;
+            }
+
+            return false;
+        }
+
+        return true;
+    }
+
+    private static function isUnionStringNotEqualOrdered(array $first_types, array $second_types): bool
+    {
+        // identical at the moment
+        return self::isUnionStringEqualOrdered($first_types, $second_types);
+    }
+
+    private static function isUnionEqualOrdered(array $first_types, array $second_types): bool {
```

Comment:
Seems like it could work, but I think we're going to need some tool to test all that. I didn't add them when it was relatively simple, but I'm not confident enough to be sure that all cases we didn't exclude are consistent now